### PR TITLE
Introduce CaptureControlInterface

### PR DESCRIPTION
**Pull request by reichlfl**
This new interface provides all functionality related to capture
control, and is required for the upcoming unit tests for
`CaptureWindow`.

Bug: http://b/224764161
Test: Compiles

### CODE REVIEW

File: src/OrbitGl/CaptureWindow.cpp [hunk 1]
```diff
@@ -72,6 +72,8 @@ class AccessibleCaptureWindow : public AccessibleWidgetBridge {
 using orbit_client_protos::TimerInfo;
 
 CaptureWindow::CaptureWindow(OrbitApp* app) : GlCanvas(), app_{app} {
+  capture_client_app_ = app;
```
Comment (vickyliu-go4it) [inline src/OrbitGl/CaptureWindow.cpp:75]: Nit: this could also be initialised in the initializer list.

File: src/OrbitGl/App.h [hunk 2]
```diff
@@ -114,14 +116,21 @@ class OrbitApp final : public DataViewFactory,
...
+  void StartCapture() override;
+  void StopCapture() override;
+  void AbortCapture() override;
+  void ToggleCapture() override;
+  // ^^^
+
   void ClearCapture();
+
+  [[nodiscard]] bool IsLoadingCapture() const;
+  [[nodiscard]] bool IsDataStreamingActive() const;
```
Comment (vickyliu-go4it) [inline src/OrbitGl/App.h:132]: override

File: src/OrbitGl/App.h [hunk 3]
```diff
@@ -114,14 +116,21 @@ class OrbitApp final : public DataViewFactory,
                                                            const std::filesystem::path& dest);
   void OnLoadCaptureCancelRequested();
 
-  [[nodiscard]] orbit_capture_client::CaptureClient::State GetCaptureState() const;
+  // vvv AppCaptureClientInterface
```
Comment (vickyliu-go4it) [inline src/OrbitGl/App.h:119]: I guess this will be removed later?

File: src/CaptureClient/include/CaptureClient/AppInterface.h [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+class AppInterface {
+ public:
+  virtual ~AppInterface() = default;
+
+  [[nodiscard]] virtual CaptureClient::State GetCaptureState() const = 0;
+  [[nodiscard]] virtual bool IsCapturing() const = 0;
+
+  virtual void StartCapture() = 0;
+  virtual void StopCapture() = 0;
+  virtual void AbortCapture() = 0;
```
Comment (vickyliu-go4it) [inline src/CaptureClient/include/CaptureClient/AppInterface.h:21]: Are these methods needed in the `CaptureWindow` unit tests? But it seems that they are invoked from `CaptureWindow`.

File: src/OrbitGl/App.cpp [hunk 5]
```diff
@@ -2661,6 +2661,8 @@ bool OrbitApp::IsLoadingCapture() const {
   return data_source_ == orbit_client_data::CaptureData::DataSource::kLoadedCapture;
 }
 
+bool OrbitApp::IsDataStreamingActive() const { return IsCapturing(); }
```
Comment (vickyliu-go4it) [inline src/OrbitGl/App.cpp:2664]: This reminds me that we might have methods shared by different modules. If we consider defining different `AppInterface`s for different modules, should we extract those shared funtionality into a `CommonAppInterface`?

File: src/DataViews/include/DataViews/AppInterface.h [hunk 6]
```diff
@@ -51,7 +51,7 @@ class AppInterface {
   virtual void SetHighlightedFunctionId(uint64_t highlighted_function_id) = 0;
   virtual void SetVisibleFunctionIds(absl::flat_hash_set<uint64_t> visible_functions) = 0;
   virtual void DeselectTimer() = 0;
-  [[nodiscard]] virtual bool IsCapturing() const = 0;
+  [[nodiscard]] virtual bool IsDataStreamingActive() const = 0;
```
Comment (beckerhe) [inline src/DataViews/include/DataViews/AppInterface.h:54]: This change seems to be rather unrelated. Is there a reason for renaming the function?

Comment (beckerhe) [reply src/DataViews/include/DataViews/AppInterface.h:54]: Ah, is it about having it in both interfaces? Why not the same name?

File: src/OrbitGl/CaptureWindow.cpp (hunk 1 above)
Comment (beckerhe) [reply src/OrbitGl/CaptureWindow.cpp:75]: I would also prefer to have it in the member initializer list. BTW, is `app_` still needed?

File: src/OrbitGl/App.h (hunk 3 above)
Comment (beckerhe) [reply src/OrbitGl/App.h:119]: I think this is supposed to be a section comment. (See the other comment below `// ^^^`)

Since this was confusing to at least one reviewer and since we don't use this style anywhere else in the code base, maybe we should either remove it or replace it by dashes `// ----------`? WDYT?

I'm undecided. Of course we shouldn't need sections because our classes are small and have only one purpose. :smile: 
But it's also clear we won't fix that in this PR. :innocent:

**Review by beckerhe — APPROVED** [review]
Otherwise LGTM

File: src/OrbitGl/CaptureWindow.cpp (hunk 1 above)
Comment (reichlfl) [reply src/OrbitGl/CaptureWindow.cpp:75]: Yes, currently we need both - I wanted to emphasize a reasonable scope and responsibility for the interface instead of simply including everything needed in the capture window. There's still parts of `CaptureWindow` that will rely on `app_ != nullptr` and can't be called in unit tests for the time being, this needs to be improved in the future.

File: src/OrbitGl/App.h (hunk 3 above)
Comment (reichlfl) [reply src/OrbitGl/App.h:119]: 😅 I agree, I'm not a fan of the sections either, but it's also confusing to have this large class implementing multiple interfaces without any grouping... I'll replace it with dashes for the time being.

File: src/CaptureClient/include/CaptureClient/AppInterface.h (hunk 4 above)
Comment (reichlfl) [reply src/CaptureClient/include/CaptureClient/AppInterface.h:21]: I thought they were, but now that you mention it - only `ToggleCapture` is used. But as I commented above, I wanted to create an interface with a somewhat narrow, but complete scope ("Control the capture") and not only limit it to how `CaptureWindow` is interacting with OrbitApp. Does that make sense to you?

File: src/DataViews/include/DataViews/AppInterface.h (hunk 6 above)
Comment (reichlfl) [reply src/DataViews/include/DataViews/AppInterface.h:54]: Yes, it's about having it in both interfaces. I was assuming the same name would cause issues, I didn't even consider trying it to be honest...

File: src/CaptureClient/include/CaptureClient/AppInterface.h (hunk 4 above)
Comment (vickyliu-go4it) [reply src/CaptureClient/include/CaptureClient/AppInterface.h:21]: SGTM! Thanks for the explanation!

File: src/DataViews/include/DataViews/AppInterface.h (hunk 6 above)
Comment (vickyliu-go4it) [reply src/DataViews/include/DataViews/AppInterface.h:54]: Woo! I also didn't know they can have the same! Good to know that XD

**Review by vickyliu-go4it — APPROVED** [review]

File: src/OrbitGl/App.cpp (hunk 5 above)
Comment (reichlfl) [reply src/OrbitGl/App.cpp]: I've now just named them the same in both interfaces. I think that in the long run, the DataViews should not define interfaces, but rather rely on the ones defined in other places... but for now, I'd just leave it as is.

File: src/DataViews/include/DataViews/AppInterface.h (hunk 6 above)
Comment (reichlfl) [reply src/DataViews/include/DataViews/AppInterface.h]: TIL that this works. That's better, thanks.

File: src/OrbitGl/App.cpp (hunk 5 above)
Comment (beckerhe) [reply src/OrbitGl/App.cpp]: I agree 100%. This is definitely a temporary state. DataViews (or later ItemModels) should get its data via Setters and should send events using signals. No coupling or interfaces required apart from the item data type.
